devmand/gateway/src/devmand/channels/cli/datastore/DatastoreTransaction.cpp

```diff
@@ -0,0 +1,361 @@
+      return DatastoreDiffType::create;
+    case LLLYD_DIFF_END:
+      throw DatastoreException("This diff type is not supported");
+    case LLLYD_DIFF_MOVEDAFTER1:
+      throw DatastoreException("This diff type is not supported");
+    case LLLYD_DIFF_MOVEDAFTER2:
+      throw DatastoreException("This diff type is not supported");
+    default:
+      throw DatastoreException("This diff type is not supported");
+  }
+}
+
+string DatastoreTransaction::buildFullPath(lllyd_node* node, string pathSoFar) {
+  string path("/");
+  path.append(node->schema->module->name)
+      .append(":")
+      .append(node->schema->name)
+      .append(pathSoFar);
+  if (node->parent == nullptr) {
+    return path;
+  }
+  return buildFullPath(node->parent, path);
+}
+void DatastoreTransaction::printDiffType(LLLYD_DIFFTYPE type) {
+  switch (type) {
+    case LLLYD_DIFF_DELETED:
+      MLOG(MINFO) << "deleted subtree:";
+      break;
+    case LLLYD_DIFF_CHANGED:
+      MLOG(MINFO) << "changed value:";
+      break;
+    case LLLYD_DIFF_MOVEDAFTER1:
+      MLOG(MINFO) << "subtree was moved one way:";
+      break;
+    case LLLYD_DIFF_MOVEDAFTER2:
+      MLOG(MINFO) << "subtree was moved another way:";
+      break;
+    case LLLYD_DIFF_CREATED:
+      MLOG(MINFO) << "subtree was added:";
+      break;
+    case LLLYD_DIFF_END:
+      MLOG(MINFO) << "end of diff:";
+  }
+}
+
+dynamic DatastoreTransaction::read(Path path) {
+  checkIfCommitted();
+
+  llly_set* pSet = lllyd_find_path(root, const_cast<char*>(path.str().c_str()));
+
+  if (pSet == nullptr) {
+    return nullptr;
+  }
+
+  if (pSet->number == 0) {
+    llly_set_free(pSet);
+    return nullptr;
+  }
+
+  if (pSet->number > 1) {
+    llly_set_free(pSet);
+
+    DatastoreException ex(
+        "Too many results from path: " + path.str() +
+        " , query must target a unique element");
+    MLOG(MWARNING) << ex.what();
+    throw ex;
+  }
+
+  const string& json = toJson(pSet->set.d[0]);
+  llly_set_free(pSet);
+  return parseJson(json);
+}
+
+void DatastoreTransaction::print(LeafVector& v) {
+  for (const auto& item : v) {
+    MLOG(MINFO) << "full path: " << item.first << " data: " << item.second;
+  }
+}
+
+lllyd_node* DatastoreTransaction::getExistingNode(
+    lllyd_node* a,
+    lllyd_node* b,
+    DatastoreDiffType type) {
+  if (type == DatastoreDiffType::create && b != nullptr) {
+    return b;
+  }
+  if (type == DatastoreDiffType::deleted && a != nullptr) {
+    return a;
+  }
+
+  return a == nullptr ? b : a;
+}
+
+bool DatastoreTransaction::isValid() {
+  checkIfCommitted();
+  if (root == nullptr) {
+    DatastoreException ex(
+        "Datastore is empty and no changes performed, nothing to validate");
+    MLOG(MWARNING) << ex.what();
+    throw ex;
+  }
+  return lllyd_validate(&root, datastoreTypeToLydOption(), nullptr) == 0;
+}
+
+int DatastoreTransaction::datastoreTypeToLydOption() {
+  switch (datastoreState->type) {
+    case operational:
+      return LLLYD_OPT_GET;
```

Comment:
comments explaining validation attached to these options and why would be useful